### DISCUSSION TITLE
Calculating degree for GCN

### DISCUSSION THREAD

**Discussion by gau-nernst** [pyg-team/pytorch_geometric#5100] [Q&A]
According to the paragraph after Eq. 2 in the GCN paper (https://arxiv.org/pdf/1609.02907.pdf), degree is calculated by summing each row in the adjacency matrix.

$$ \tilde D_{ii} = \sum_j{\tilde A_{ij}} $$

I believe this is equivalent to out-degree for directed graphs.

Currently `gcn_norm()` calculates degree by summing each column (in-degree) instead:

https://github.com/pyg-team/pytorch_geometric/blob/4f392ace20725b8b8895a618bea4d728439933fb/torch_geometric/nn/conv/gcn_conv.py#L63-L64

Summing each row or summing each column are the same for undirected graphs. I notice that before commit 72e8ef33, degree is summed for each row. Although the original GCN paper is not meant for directed graphs, what is the reason for this change?

Right now, to get the original behaviour of GCN on directed graphs (which aggregates target to source), I can pass `flow="target_to_source"` in `GCNConv`, but I cannot change the way degree is calculated when normalizing the adjacency matrix. Would it make sense to add an option to select how degree is calculated? (in-degree or out-degree, which will pass this parameter to `gcn_norm()`)

**Comment by wsad1** [comment]
When `flow="source_to_target"`, the in-degree represents the number of incoming messages, so it makes sense to normalize by in-degree and not out-degree.
However, when the flow is reversed we should be computing the norm using out-degree  `deg = scatter_add(edge_weight, row, dim=0, dim_size=num_nodes) `.  Thanks for bringing this up.

I don't think we need an option to control how degree is computed it should depend only on the `flow`.

**Reply by rusty1s** [reply]
Yes, I think we should fix this by passing `flow` to `gcn_norm`. @wsad1 @gau-nernst Do you have interest in fixing this?

**Reply by gau-nernst** [reply]
@rusty1s Sure I can prepare a PR for this. I think the doc will need to be updated to reflect this behavior also. Do you want to add a unit test for this?

**Reply by rusty1s** [reply]
This would be amazing, thank so you much! We can add a test that confirms that
```python
conv.flow = 'source_to_target'
out1 = conv(x, edge_index)
conv.flow = 'target_to_source'
out2 = conv(x, edge_index.flip([0])
assert torch.allclose(out1, out2)
```
